automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala

```diff
@@ -1,56 +1,155 @@
+
+    cluster.googleProject shouldBe project
+    cluster.clusterName shouldBe expectedName
+    cluster.googleServiceAccount shouldBe sa
+    cluster.googleBucket shouldBe bucket
+    cluster.labels shouldBe requestedLabels ++ DefaultLabels(expectedName, project, bucket, sa, None).toMap
+    expectedStatuses should contain (cluster.status)
+  }
+
+  // creates a cluster and checks to see that it reaches the Running state
+  def createAndMonitor(googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest): Cluster = {
+    val cluster = Leonardo.cluster.create(googleProject, clusterName, clusterRequest)
+    clusterCheck(cluster, clusterRequest.labels, clusterName, Seq(ClusterStatus.Creating))
+
+    // verify with get()
+    clusterCheck(Leonardo.cluster.get(googleProject, clusterName), clusterRequest.labels, clusterName, Seq(ClusterStatus.Creating))
+
+    // wait for "Running" or error (fail fast)
+    eventually {
+      clusterCheck(Leonardo.cluster.get(googleProject, clusterName), clusterRequest.labels, clusterName, Seq(ClusterStatus.Running, ClusterStatus.Error))
+    } (PatienceConfig(timeout = scaled(Span(6, Minutes)), interval = scaled(Span(20, Seconds))))
+
+    val runningCluster = Leonardo.cluster.get(googleProject, clusterName)
+    clusterCheck(runningCluster, clusterRequest.labels, clusterName, Seq(ClusterStatus.Running))
+    runningCluster
+  }
+
+  // TODO: show diffs as screenshot or other test output?
+  def compareFilesExcludingIPs(left: File, right: File): Unit = {
+
+    def linesWithoutIPs(file: File) = {
+      import scala.collection.JavaConverters._
+      Files.readAllLines(left.toPath).asScala map { _.replaceAll("(\\d+.){3}\\d+", "<IP>") }
+    }
+
+    linesWithoutIPs(left) shouldEqual linesWithoutIPs(right)
+  }
+
+  // ------------------------------------------------------
+
+  def withNewCluster[T](googleProject: GoogleProject)(testCode: Cluster => Any): Any = {
+    val name = ClusterName(s"automation-test-a${Random.alphanumeric.take(10).mkString.toLowerCase}z")
+    val request = ClusterRequest(bucket, sa, Map("foo" -> Random.alphanumeric.take(10).mkString), None)
+
+    try {
+      val cluster = createAndMonitor(googleProject, name, request)
+      val testResult = testCode(cluster)
+      testResult
+    }
+    finally {
+      Leonardo.cluster.delete(googleProject, name) shouldBe "The request has been accepted for processing, but the processing has not been completed."
+    }
+  }
+
+  def withNotebooksListPage[T](cluster: Cluster)(testCode: NotebooksListPage => T)(implicit webDriver: WebDriver): T = {
+    // GAWB-2797: Leo might not be ready to proxy yet
+    Thread sleep 1000 * 10
+
+    val notebooksListPage = Leonardo.notebooks.get(cluster.googleProject, cluster.clusterName)
+    testCode(notebooksListPage.open)
+  }
+
+  def withNotebookUpload[T](cluster: Cluster, file: File)(testCode: NotebookPage => T)(implicit webDriver: WebDriver): T = {
```

Comment:
This does not delete the uploaded notebook from the cluster, because we always delete the cluster